### ISSUE TITLE
Inconsistent delay between retries to non-listening ports

### ISSUE THREAD

**Issue by kevinAlbs** [c-ares/c-ares#819]
Running [this example](https://gist.github.com/kevinAlbs/e5514247c1b7f6a688d9561e53392742) sends an A query to `localhost:12345`. With no process listening on port 12345, the queries appear to fail and retry immediately. Here is a capture with Wireshark:

![ipv4-retries](https://github.com/user-attachments/assets/b606c25b-c7ed-4276-91c4-0325323f50a9)

However, running the same example with `AF_UNSPEC` (set the environment variable `USE_AF_UNSPEC` in the example) results in an appeared delayed retry of the A query by 3 seconds:

![ipv4-and-ipv4-retries](https://github.com/user-attachments/assets/3f35d0d1-4e5c-466a-b359-0588f0cf321d)

I expected both retry behaviors to be similar.

**Background**

A user reported seeing ECONNREFUSED errors from the MongoDB Node driver during start-up of a Kubernetes cluster. Wireshark captures suggest DNS resolution is retrying immediately and failing on retry. I expect this is due to a delay in the DNS server starting (waiting until after the DNS server starts to start application processes appears to resolve the issue). This led me to investigate the expected DNS retry behavior.

**Comment by bradh352** [comment]
I would need to know the version of c-ares in use.  There has been a lot of changes over the last few releases.

**Comment by kevinAlbs** [comment]
The example was using c-ares built on commit bf7c946b769e722c3e3e4062b24393ede1b3bb51. Modifying the example to print `ares_version` shows `1.32.2`.

**Comment by bradh352** [comment]
What OS/Version are you on?

At least on my MacOS system, there's no delay.  My guess is the ICMP Destination Unreachable packet on your system is being dropped by the OS so c-ares has no way to know the UDP session is invalid.  UDP is connectionless, so short of the helpful ICMP reply in this case (which really just so happens to come back on loopback), you've really got to rely on retries.

AF_INET
```
sudo tcpdump -i lo0 -vv -X
tcpdump: listening on lo0, link-type NULL (BSD loopback), snapshot length 524288 bytes
11:29:41.638330 IP (tos 0x0, ttl 64, id 19286, offset 0, flags [none], proto UDP (17), length 68, bad cksum 0 (->3151)!)
    localhost.62385 > localhost.italk: [bad udp cksum 0xfe43 -> 0xeadc!] UDP, length 40
	0x0000:  4500 0044 4b56 0000 4011 0000 7f00 0001  E..DKV..@.......
	0x0010:  7f00 0001 f3b1 3039 0030 fe43 1e5c 0100  ......09.0.C.\..
	0x0020:  0001 0000 0000 0001 0765 7861 6d70 6c65  .........example
	0x0030:  0363 6f6d 0000 0100 0100 0029 04d0 0000  .com.......)....
	0x0040:  0000 0000                                ....
11:29:41.638424 IP (tos 0x0, ttl 64, id 52640, offset 0, flags [none], proto ICMP (1), length 56, bad cksum 0 (->af22)!)
    localhost > localhost: ICMP localhost udp port italk unreachable, length 36
	IP (tos 0x0, ttl 64, id 19286, offset 0, flags [none], proto UDP (17), length 68, bad cksum 0 (->3151)!)
    localhost.62385 > localhost.italk: [no cksum] UDP, length 40
	0x0000:  4500 0038 cda0 0000 4001 0000 7f00 0001  E..8....@.......
	0x0010:  7f00 0001 0303 0a33 0000 0000 4500 0044  .......3....E..D
	0x0020:  4b56 0000 4011 0000 7f00 0001 7f00 0001  KV..@...........
	0x0030:  f3b1 3039 0030 0000                      ..09.0..
11:29:41.638531 IP (tos 0x0, ttl 64, id 14476, offset 0, flags [none], proto UDP (17), length 68, bad cksum 0 (->441b)!)
    localhost.57282 > localhost.italk: [bad udp cksum 0xfe43 -> 0xfecb!] UDP, length 40
	0x0000:  4500 0044 388c 0000 4011 0000 7f00 0001  E..D8...@.......
	0x0010:  7f00 0001 dfc2 3039 0030 fe43 1e5c 0100  ......09.0.C.\..
	0x0020:  0001 0000 0000 0001 0765 7861 6d70 6c65  .........example
	0x0030:  0363 6f6d 0000 0100 0100 0029 04d0 0000  .com.......)....
	0x0040:  0000 0000                                ....
11:29:41.638562 IP (tos 0x0, ttl 64, id 22920, offset 0, flags [none], proto ICMP (1), length 56, bad cksum 0 (->233b)!)
    localhost > localhost: ICMP localhost udp port italk unreachable, length 36
	IP (tos 0x0, ttl 64, id 14476, offset 0, flags [none], proto UDP (17), length 68, bad cksum 0 (->441b)!)
    localhost.57282 > localhost.italk: [no cksum] UDP, length 40
	0x0000:  4500 0038 5988 0000 4001 0000 7f00 0001  E..8Y...@.......
	0x0010:  7f00 0001 0303 30ec 0000 0000 4500 0044  ......0.....E..D
	0x0020:  388c 0000 4011 0000 7f00 0001 7f00 0001  8...@...........
	0x0030:  dfc2 3039 0030 0000                      ..09.0..
11:29:41.638626 IP (tos 0x0, ttl 64, id 65430, offset 0, flags [none], proto UDP (17), length 68, bad cksum 0 (->7d10)!)
    localhost.50276 > localhost.italk: [bad udp cksum 0xfe43 -> 0x1a2a!] UDP, length 40
	0x0000:  4500 0044 ff96 0000 4011 0000 7f00 0001  E..D....@.......
	0x0010:  7f00 0001 c464 3039 0030 fe43 1e5c 0100  .....d09.0.C.\..
	0x0020:  0001 0000 0000 0001 0765 7861 6d70 6c65  .........example
	0x0030:  0363 6f6d 0000 0100 0100 0029 04d0 0000  .com.......)....
	0x0040:  0000 0000                                ....
11:29:41.638657 IP (tos 0x0, ttl 64, id 13882, offset 0, flags [none], proto ICMP (1), length 56, bad cksum 0 (->4689)!)
    localhost > localhost: ICMP localhost udp port italk unreachable, length 36
	IP (tos 0x0, ttl 64, id 65430, offset 0, flags [none], proto UDP (17), length 68, bad cksum 0 (->7d10)!)
    localhost.50276 > localhost.italk: [no cksum] UDP, length 40
	0x0000:  4500 0038 363a 0000 4001 0000 7f00 0001  E..86:..@.......
	0x0010:  7f00 0001 0303 853f 0000 0000 4500 0044  .......?....E..D
	0x0020:  ff96 0000 4011 0000 7f00 0001 7f00 0001  ....@...........
	0x0030:  c464 3039 0030 0000                      .d09.0..
```

AF_UNSPEC:
```
$ sudo tcpdump -i lo0 -vv -X
tcpdump: listening on lo0, link-type NULL (BSD loopback), snapshot length 524288 bytes
11:30:27.761639 IP (tos 0x0, ttl 64, id 17696, offset 0, flags [none], proto UDP (17), length 68, bad cksum 0 (->3787)!)
    localhost.63810 > localhost.italk: [bad udp cksum 0xfe43 -> 0xdab5!] UDP, length 40
	0x0000:  4500 0044 4520 0000 4011 0000 7f00 0001  E..DE...@.......
	0x0010:  7f00 0001 f942 3039 0030 fe43 28f2 0100  .....B09.0.C(...
	0x0020:  0001 0000 0000 0001 0765 7861 6d70 6c65  .........example
	0x0030:  0363 6f6d 0000 0100 0100 0029 04d0 0000  .com.......)....
	0x0040:  0000 0000                                ....
11:30:27.761675 IP (tos 0x0, ttl 64, id 828, offset 0, flags [none], proto UDP (17), length 68, bad cksum 0 (->796b)!)
    localhost.63810 > localhost.italk: [bad udp cksum 0xfe43 -> 0x38d1!] UDP, length 40
	0x0000:  4500 0044 033c 0000 4011 0000 7f00 0001  E..D.<..@.......
	0x0010:  7f00 0001 f942 3039 0030 fe43 afd6 0100  .....B09.0.C....
	0x0020:  0001 0000 0000 0001 0765 7861 6d70 6c65  .........example
	0x0030:  0363 6f6d 0000 1c00 0100 0029 04d0 0000  .com.......)....
	0x0040:  0000 0000                                ....
11:30:27.761726 IP (tos 0x0, ttl 64, id 45477, offset 0, flags [none], proto ICMP (1), length 56, bad cksum 0 (->cb1d)!)
    localhost > localhost: ICMP localhost udp port italk unreachable, length 36
	IP (tos 0x0, ttl 64, id 17696, offset 0, flags [none], proto UDP (17), length 68, bad cksum 0 (->3787)!)
    localhost.63810 > localhost.italk: [no cksum] UDP, length 40
	0x0000:  4500 0038 b1a5 0000 4001 0000 7f00 0001  E..8....@.......
	0x0010:  7f00 0001 0303 0ad8 0000 0000 4500 0044  ............E..D
	0x0020:  4520 0000 4011 0000 7f00 0001 7f00 0001  E...@...........
	0x0030:  f942 3039 0030 0000                      .B09.0..
11:30:27.761742 IP (tos 0x0, ttl 64, id 8800, offset 0, flags [none], proto ICMP (1), length 56, bad cksum 0 (->5a63)!)
    localhost > localhost: ICMP localhost udp port italk unreachable, length 36
	IP (tos 0x0, ttl 64, id 828, offset 0, flags [none], proto UDP (17), length 68, bad cksum 0 (->796b)!)
    localhost.63810 > localhost.italk: [no cksum] UDP, length 40
	0x0000:  4500 0038 2260 0000 4001 0000 7f00 0001  E..8"`..@.......
	0x0010:  7f00 0001 0303 4cbc 0000 0000 4500 0044  ......L.....E..D
	0x0020:  033c 0000 4011 0000 7f00 0001 7f00 0001  .<..@...........
	0x0030:  f942 3039 0030 0000                      .B09.0..
11:30:27.761863 IP (tos 0x0, ttl 64, id 26680, offset 0, flags [none], proto UDP (17), length 68, bad cksum 0 (->146f)!)
    localhost.65410 > localhost.italk: [bad udp cksum 0xfe43 -> 0xd475!] UDP, length 40
	0x0000:  4500 0044 6838 0000 4011 0000 7f00 0001  E..Dh8..@.......
	0x0010:  7f00 0001 ff82 3039 0030 fe43 28f2 0100  ......09.0.C(...
	0x0020:  0001 0000 0000 0001 0765 7861 6d70 6c65  .........example
	0x0030:  0363 6f6d 0000 0100 0100 0029 04d0 0000  .com.......)....
	0x0040:  0000 0000                                ....
11:30:27.761883 IP (tos 0x0, ttl 64, id 54554, offset 0, flags [none], proto UDP (17), length 68, bad cksum 0 (->a78c)!)
    localhost.65410 > localhost.italk: [bad udp cksum 0xfe43 -> 0x3291!] UDP, length 40
	0x0000:  4500 0044 d51a 0000 4011 0000 7f00 0001  E..D....@.......
	0x0010:  7f00 0001 ff82 3039 0030 fe43 afd6 0100  ......09.0.C....
	0x0020:  0001 0000 0000 0001 0765 7861 6d70 6c65  .........example
	0x0030:  0363 6f6d 0000 1c00 0100 0029 04d0 0000  .com.......)....
	0x0040:  0000 0000                                ....
11:30:27.761898 IP (tos 0x0, ttl 64, id 3033, offset 0, flags [none], proto ICMP (1), length 56, bad cksum 0 (->70ea)!)
    localhost > localhost: ICMP localhost udp port italk unreachable, length 36
	IP (tos 0x0, ttl 64, id 26680, offset 0, flags [none], proto UDP (17), length 68, bad cksum 0 (->146f)!)
    localhost.65410 > localhost.italk: [no cksum] UDP, length 40
	0x0000:  4500 0038 0bd9 0000 4001 0000 7f00 0001  E..8....@.......
	0x0010:  7f00 0001 0303 e17f 0000 0000 4500 0044  ............E..D
	0x0020:  6838 0000 4011 0000 7f00 0001 7f00 0001  h8..@...........
	0x0030:  ff82 3039 0030 0000                      ..09.0..
11:30:27.761909 IP (tos 0x0, ttl 64, id 234, offset 0, flags [none], proto ICMP (1), length 56, bad cksum 0 (->7bd9)!)
    localhost > localhost: ICMP localhost udp port italk unreachable, length 36
	IP (tos 0x0, ttl 64, id 54554, offset 0, flags [none], proto UDP (17), length 68, bad cksum 0 (->a78c)!)
    localhost.65410 > localhost.italk: [no cksum] UDP, length 40
	0x0000:  4500 0038 00ea 0000 4001 0000 7f00 0001  E..8....@.......
	0x0010:  7f00 0001 0303 749d 0000 0000 4500 0044  ......t.....E..D
	0x0020:  d51a 0000 4011 0000 7f00 0001 7f00 0001  ....@...........
	0x0030:  ff82 3039 0030 0000                      ..09.0..
11:30:27.761965 IP (tos 0x0, ttl 64, id 64703, offset 0, flags [none], proto UDP (17), length 68, bad cksum 0 (->7fe7)!)
    localhost.51349 > localhost.italk: [bad udp cksum 0xfe43 -> 0x0b63!] UDP, length 40
	0x0000:  4500 0044 fcbf 0000 4011 0000 7f00 0001  E..D....@.......
	0x0010:  7f00 0001 c895 3039 0030 fe43 28f2 0100  ......09.0.C(...
	0x0020:  0001 0000 0000 0001 0765 7861 6d70 6c65  .........example
	0x0030:  0363 6f6d 0000 0100 0100 0029 04d0 0000  .com.......)....
	0x0040:  0000 0000                                ....
11:30:27.761983 IP (tos 0x0, ttl 64, id 14381, offset 0, flags [none], proto UDP (17), length 68, bad cksum 0 (->447a)!)
    localhost.51349 > localhost.italk: [bad udp cksum 0xfe43 -> 0x697e!] UDP, length 40
	0x0000:  4500 0044 382d 0000 4011 0000 7f00 0001  E..D8-..@.......
	0x0010:  7f00 0001 c895 3039 0030 fe43 afd6 0100  ......09.0.C....
	0x0020:  0001 0000 0000 0001 0765 7861 6d70 6c65  .........example
	0x0030:  0363 6f6d 0000 1c00 0100 0029 04d0 0000  .com.......)....
	0x0040:  0000 0000                                ....
11:30:27.761995 IP (tos 0x0, ttl 64, id 45387, offset 0, flags [none], proto ICMP (1), length 56, bad cksum 0 (->cb77)!)
    localhost > localhost: ICMP localhost udp port italk unreachable, length 36
	IP (tos 0x0, ttl 64, id 64703, offset 0, flags [none], proto UDP (17), length 68, bad cksum 0 (->7fe7)!)
    localhost.51349 > localhost.italk: [no cksum] UDP, length 40
	0x0000:  4500 0038 b14b 0000 4001 0000 7f00 0001  E..8.K..@.......
	0x0010:  7f00 0001 0303 83e5 0000 0000 4500 0044  ............E..D
	0x0020:  fcbf 0000 4011 0000 7f00 0001 7f00 0001  ....@...........
	0x0030:  c895 3039 0030 0000                      ..09.0..
11:30:27.762006 IP (tos 0x0, ttl 64, id 65168, offset 0, flags [none], proto ICMP (1), length 56, bad cksum 0 (->7e32)!)
    localhost > localhost: ICMP localhost udp port italk unreachable, length 36
	IP (tos 0x0, ttl 64, id 14381, offset 0, flags [none], proto UDP (17), length 68, bad cksum 0 (->447a)!)
    localhost.51349 > localhost.italk: [no cksum] UDP, length 40
	0x0000:  4500 0038 fe90 0000 4001 0000 7f00 0001  E..8....@.......
	0x0010:  7f00 0001 0303 4878 0000 0000 4500 0044  ......Hx....E..D
	0x0020:  382d 0000 4011 0000 7f00 0001 7f00 0001  8-..@...........
	0x0030:  c895 3039 0030 0000                      ..09.0..
```

**Comment by bradh352** [comment]
I just tested on linux and can see it happen there.  looks like on a send() one query is successful, and the other gets an ECONNREFUSED errno (due to the ICMP unreachable having propagated back by this point from the first query), so this causes the second query to get requeued immediately, but the first just hangs out until the timeout is hit.

On MacOS, however, we get an event that causes us to recv() then we catch the error and clean up the udp session and requeue any packets associated with the session.  I'm assuming Linux doesn't propagate another event because it assumes we did something when send() failed.

Looks like what we need to do is to look at the reason for the send() error and determine if we need to do the same udp session cleanup that a recv() error would do.
